internal/check/engine.go

```diff
@@ -47,14 +50,24 @@ func (e *Engine) subjectIsAllowed(ctx context.Context, requested *relationtuple.
 			continue
 		}
 
-		var err error
+		nm, err := e.d.NamespaceManager()
+		if err != nil {
+			return false, err
+		}
+		n, err := nm.GetNamespace(ctx, sr.Namespace)
+		if err != nil {
+			return false, err
+		}
+		// use the namespace here to decide which bool operators to use in the loop, and how to adjust the query when moving on (e.g. different relation name)
+		var strategy func(ctx context.Context, requested, current *relationtuple.InternalRelationTuple, checkFurther func(context.Context, *relationtuple.InternalRelationTuple, *relationtuple.RelationQuery) (bool, error)) (continueLoop, allowed bool, err error) = n.GetRelationRewriteStrategy(ctx, sr.Relation)
```

Comment:
What does `n.GetRelationRewriteStrategy` do? I don't see that defined in the `namespace/definitions.go`

Reply:
That is yet to be implemented. In case of a CEL like rewrites it would probably just evaluate the expression.
In case of union the strategy returns true iff one of the subject sets evaluates to true, in case of intersection all would have to, and in case of exclusion it has to evaluate to false.

Until now I returned true at this point iff I found a subject set that returned true. The strategy just implements the logic as required by the config.

Reply:
Ok, that makes sense at a high level. I'm curious to see that in practice. The challenge I foresee is the nested sub-expressions. Each rewrite expression can involve subexpressions, so these rewrite "strategies" could nest quite deep.

Reply:
Sure, that is true. But I guess the strategies will be quite simple, at least for now. I am sure that we can optimize it later on. Having the graph search controlled by these strategies will probably be the easiest and most flexible approach.